src/turnkeyml/build/stage.py

```diff
@@ -342,11 +350,6 @@ def launch(self, state: build.State) -> build.State:
 
             return state
 
-        finally:
-            # Collect telemetry about the stage
-            execution_time = time.time() - start_time
-            stats.save_model_eval_stat(stage.duration_key, execution_time)
```

Comment:
I see what the bug was 🤣 

Still, I kinda like that `finally` block because we still get the stage duration even when there is an exception.

What about moving the entire try-except-finally block inside of the `for stage in self.stages:` loop? Would that fix the bug while still giving us stage durations on failed stages?

Reply:
The implementation you reviewed seemed to indeed have a problem: We would not record the stage duration if an exception was raised AND that exception was not a stageError (Errors we failed to anticipate).

The revised implementation ensures that we keep track of the duration regardless of the exception raised.

Reply:
Alternatively, here is another implementation that should work exactly the same way, but is a bit less verbose, easier to understand, and also passes the `cli.py` CI test. Let me know which one you prefer:

```
def launch(self, state: build.State) -> build.State:

        ...SAME CODE AS BEFORE...
        
        # Run the build
        for stage in self.stages:
            try:
                start_time = time.time()

                # Set status as incomplete, since stage just started
                stats.save_model_eval_stat(
                    stage.status_key, build.FunctionStatus.INCOMPLETE.value
                )

                # Collect telemetry about the stage
                state.current_build_stage = stage.unique_name

                # Run the stage
                state = stage.fire_helper(state)

            # Broad exception is desirable as we want to capture
            # all exceptions (including those we can't anticipate)
            except Exception as e: # pylint: disable=broad-except

                # Update Stage Status
                stats.save_model_eval_stat(
                    stage.status_key, build.FunctionStatus.ERROR.value
                )

                # Advance the cursor below the monitor so
                # we can print an error message
                stage_depth_in_sequence = self.get_depth() - self.get_names().index(
                    stage.unique_name  # pylint: disable=undefined-loop-variable
                )
                stdout_lines_to_advance = stage_depth_in_sequence - 2
                cursor_down = "\n" * stdout_lines_to_advance

                print(cursor_down)

                printing.log_error(e)

                raise

            else:
                # Update Stage Status
                stats.save_model_eval_stat(
                    stage.status_key, build.FunctionStatus.SUCCESSFUL.value
                )

            finally:
                # Store stage duration
                execution_time = time.time() - start_time
                stats.save_model_eval_stat(stage.duration_key, execution_time)

        state.current_build_stage = None
        state.build_status = build.FunctionStatus.SUCCESSFUL

        # We use a deepcopy here because the Stage framework supports
        # intermediate_results of any type, including model objects in memory.
        # The deepcopy ensures that we are providing a result that users
        # are free to take any action with.
        state.results = copy.deepcopy(state.intermediate_results)

        return state
```

Reply:
> The implementation you reviewed seemed to indeed have a problem: We would not record the stage duration if an exception was raised AND that exception was not a stageError (Errors we failed to anticipate would not be stageErrros).

Are you sure? Here's the definition of `finally`:

> finally block is always executed after leaving the try statement. In case if some exception was not handled by except block, it is re-raised after execution of finally block.

Reply:
I like the implementation in [your comment](https://github.com/onnx/turnkeyml/pull/97/files#r1477628210) a lot, please do that!

Reply:
Done!